src/analysis/retail/priest/discipline/modules/spells/PowerWordShield.tsx

```diff
@@ -0,0 +1,181 @@
+  healing: number;
+  crit: boolean;
+};
+
+class PowerWordShield extends Analyzer {
+  static dependencies = {
+    statTracker: StatTracker,
+  };
+
+  protected statTracker!: StatTracker;
+
+  hasAegis = false;
+  decayedShields = 0;
+  shieldApplications: ShieldInfo[] = [];
+  shieldOfAbsolutionValue = 0;
+  critCount = 0;
+  pwsValue = 0;
+  aegisValue = 0;
+  t23pValue = 0;
+
+  constructor(options: Options) {
+    super(options);
+
+    this.hasAegis = this.selectedCombatant.hasTalent(TALENTS_PRIEST.AEGIS_OF_WRATH_TALENT);
+
+    //
+
+    this.addEventListener(
+      Events.applybuff.by(SELECTED_PLAYER).spell(SPELLS.POWER_WORD_SHIELD),
+      this.onShieldApplication,
+    );
+
+    // when Power Word Shield absorbs damage
+    this.addEventListener(
+      Events.absorbed.by(SELECTED_PLAYER).spell(SPELLS.POWER_WORD_SHIELD),
+      this.onPWSAbsorb,
+    );
+
+    // these will determine value to amount of the 4p buff
+    this.addEventListener(
+      Events.heal.by(SELECTED_PLAYER).spell(SPELLS.PENANCE_HEAL),
+      this.onPenanceHeal,
+    );
+    this.addEventListener(
+      Events.damage.by(SELECTED_PLAYER).spell(SPELLS.PENANCE),
+      this.onPenanceDamage,
+    );
+    this.addEventListener(
+      Events.cast.by(SELECTED_PLAYER).spell(SPELLS.PENANCE_CAST),
+      this.onPenanceCast,
+    );
+
+    this.addEventListener(
+      Events.removebuff.by(SELECTED_PLAYER).spell(SPELLS.POWER_WORD_SHIELD),
+      this.onShieldExpiry,
+    );
+  }
+
+  onShieldApplication(event: ApplyBuffEvent) {
+    this.shieldApplications.push({
+      event: event,
+      shieldOfAbsolutionValue: this.shieldOfAbsolutionValue,
+      healing: 0,
+      crit: this.critCheck(event),
+    });
+    this.shieldOfAbsolutionValue = 0;
+  }
+
+  critCheck(event: ApplyBuffEvent) {
+    // We need to check if the Shield was a crit, as the 4p value is then doubled in value. It's not possible to tell this from regular events,
+    // however when checking for rapture and using the current intellect, we can accurately determine this. The only times this will be inaccurate
+    // is when the target has a healing increase which also increases potency of absorbs (there are not many, the main offender is Vampiric blood)
+
+    let baseShield = event.absorb || 0;
+    const hasRapture = this.selectedCombatant.hasBuff(TALENTS_PRIEST.RAPTURE_TALENT.id); // Rapture amps the four piece bonus
+    baseShield /= hasRapture ? 1.3 : 1;
+    baseShield -= this.shieldOfAbsolutionValue;
+    const intellect = this.statTracker.currentIntellectRating;
+    const crit = baseShield > intellect * 10 ? true : false;
+    return crit;
+  }
+
+  onShieldExpiry(event: RemoveBuffEvent) {
+    this.shieldApplications.forEach((shieldApplication) => {
+      if (
+        event.targetID === shieldApplication.event.targetID &&
+        event.timestamp > shieldApplication.event.timestamp &&
+        shieldApplication.event.timestamp + POWER_WORD_SHIELD_DURATION_MS > event.timestamp
+      ) {
```

Comment:
for both this and `onPWSAbsorb` I would recommend converting `shieldApplications` to an object or map keyed by `targetID`. For example:

```ts
private shieldApplications: Map<number, ShieldInfo> = new Map();
```

then instead of using this foreach you do:

```ts
const info = this.shieldApplications.get(event.targetID);
if (!info || 
    info.event.timestamp > event.timestamp || // not sure how this happens? fabrication stuff?
    info.event.timestamp + POWER_WORLD_SHIELD_DURATION_MS < event.timestamp
    ) {
    return;
    }
 ```
 
 this will be significantly faster on higher group sizes / longer fights.